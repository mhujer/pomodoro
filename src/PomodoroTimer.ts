export interface PastPomodoro {
    startTime: number;
    endTime: number;
}

let POMODORO_DURATION_SECONDS: number;
// @see https://webpack.js.org/guides/production/#specify-the-mode
if (process.env.NODE_ENV === 'production') {
    POMODORO_DURATION_SECONDS = 25 * 60;
} else {
    POMODORO_DURATION_SECONDS = 5;
}

export class PomodoroTimer {

    private _startTime: number | null = null;
    private _latestNow: number;

    private _pastPomodoros: PastPomodoro[] = [];
    private defaultPomodoroDurationSeconds;

    constructor(
        defaultPomodoroDurationSeconds = POMODORO_DURATION_SECONDS
    ) {
        this.defaultPomodoroDurationSeconds = defaultPomodoroDurationSeconds;
        this._latestNow = Date.now();
    }

    startTimer() {
        this._startTime = Date.now();
        this._latestNow = Date.now();
    }

    tick() {
        if (this._startTime === null) {
            throw new Error('!');
        }

        this._latestNow = Date.now();

        const elapsedSeconds = (this._latestNow - this._startTime) / 1000;

        if (elapsedSeconds >= POMODORO_DURATION_SECONDS) {
            this._pastPomodoros.push({
                startTime: this._startTime,
                endTime: this._latestNow,
            });
            this._startTime = null;
        }
    }

    get startTime(): number | null {
        return this._startTime;
    }


    get latestNow(): number {
        return this._latestNow;
    }

    isPomodoroActive(): boolean {
        return this._startTime !== null;
    }

    getRemainingSeconds(): number|null {
        if (!this.isPomodoroActive()) {
            return null;
        }

        if (this._startTime === null) {
            throw new Error('!'); // @todo is checked by the condition above
        }

        const elapsedSeconds = (this._latestNow - this._startTime) / 1000;
        return POMODORO_DURATION_SECONDS - elapsedSeconds;
    }

    get pastPomodoros(): PastPomodoro[] {
        return [...this._pastPomodoros]; //@todo deep clone
    }

    getDefaultPomodoroDurationSeconds(): number {
        return this.defaultPomodoroDurationSeconds;
    }
}
