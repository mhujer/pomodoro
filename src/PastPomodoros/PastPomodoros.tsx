import React from 'react';
import { formatPomodoroTime } from '../timeFormatUtils';
import { lightFormat } from 'date-fns';
import { PastPomodoro } from '../PomodoroTimer';

type Props = {
    pastPomodoros: PastPomodoro[]
}
export const PastPomodoros: React.FC<Props> = ({ pastPomodoros }) => {
    // no pomodoros
    if (pastPomodoros.length === 0) {
        return null;
    }

    // @todo ✅ / ❌

    // @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description
    pastPomodoros.sort((a, b) => {
        return b.startTime - a.startTime;
    });

    const pastPomodorosLog = pastPomodoros.map((pastPomodoro) => {
        const duration = formatPomodoroTime((pastPomodoro.endTime - pastPomodoro.startTime) / 1000);
        const startDateFormatted = lightFormat(new Date(pastPomodoro.startTime), 'yyyy-MM-dd HH:mm:ss');

        return (
            <li key={pastPomodoro.startTime}>
                <em>{startDateFormatted}</em>: duration <strong>{duration}</strong>
            </li>
        );
    });

    return (
        <ul>
            {pastPomodorosLog}
        </ul>
    );
};
