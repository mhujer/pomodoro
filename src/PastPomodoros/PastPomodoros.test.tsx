import { PastPomodoros } from './PastPomodoros';
import renderer from 'react-test-renderer';
import { PastPomodoro } from '../PomodoroTimer';
import { describe, it } from 'vitest';

describe('PastPomodoros', () => {
    it('renders for no pomodoros', () => {
        const domTree = renderer.create(<PastPomodoros pastPomodoros={[]} />).toJSON();
        expect(domTree).toMatchSnapshot();
    });

    it('renders for 1 pomodoro', () => {
        const pastPomodoros: PastPomodoro[] = [
            {
                startTime: new Date('2023-02-18 15:00:00').getTime(),
                endTime: new Date('2023-02-18 15:25:00').getTime(),
                didFinish: true,
            },
        ];

        const domTree = renderer.create(<PastPomodoros pastPomodoros={pastPomodoros} />).toJSON();
        expect(domTree).toMatchSnapshot();
    });

    it('renders finished and void pomodoros', () => {
        const pastPomodoros: PastPomodoro[] = [
            {
                startTime: new Date('2023-02-18 15:00:00').getTime(),
                endTime: new Date('2023-02-18 15:05:11').getTime(),
                didFinish: false,
            },
            {
                startTime: new Date('2023-02-18 15:10:00').getTime(),
                endTime: new Date('2023-02-18 15:35:00').getTime(),
                didFinish: true,
            },
        ];

        const domTree = renderer.create(<PastPomodoros pastPomodoros={pastPomodoros} />).toJSON();
        expect(domTree).toMatchSnapshot();
    });
});
