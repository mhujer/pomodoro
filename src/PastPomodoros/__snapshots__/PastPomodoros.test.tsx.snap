// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PastPomodoros > renders finished and void pomodoros 1`] = `
<ul>
  <li>
    <em>
      2023-02-18 15:10:00
    </em>
    : duration
     
    <strong>
      25:00
       
      ✅
    </strong>
  </li>
  <li>
    <em>
      2023-02-18 15:00:00
    </em>
    : duration
     
    <strong>
      05:11
       
      ❌
    </strong>
  </li>
</ul>
`;

exports[`PastPomodoros > renders for 1 pomodoro 1`] = `
<ul>
  <li>
    <em>
      2023-02-18 15:00:00
    </em>
    : duration
     
    <strong>
      25:00
       
      ✅
    </strong>
  </li>
</ul>
`;

exports[`PastPomodoros > renders for no pomodoros 1`] = `null`;
